broker/src/main/java/org/apache/rocketmq/broker/transaction/queue/TransactionalMessageServiceImpl.java

```diff
@@ -0,0 +1,498 @@
+            log.info("Topic: {} tags: {}, OpOffset: {}, HalfOffset: {}", opMessageExt.getTopic(),
+                opMessageExt.getTags(), opMessageExt.getQueueOffset(), queueOffset);
+            if (TransactionalMessageUtil.REMOVETAG.equals(opMessageExt.getTags())) {
+                if (queueOffset < miniOffset) {
+                    doneOpOffset.add(opMessageExt.getQueueOffset());
+                } else {
+                    removeMap.put(queueOffset, opMessageExt.getQueueOffset());
+                }
+            } else {
+                log.error("Found a illegal tag in opMessageExt= {} ", opMessageExt);
+            }
+        }
+        log.debug("Remove map: {}", removeMap);
+        log.debug("Done op list: {}", doneOpOffset);
+        return pullResult;
+    }
+
+    /**
+     * If return true, skip this msg
+     *
+     * @param removeMap Op message map to determine whether a half message was responded by producer.
+     * @param doneOpOffset Op Message which has been checked.
+     * @param msgExt Half message
+     * @param checkImmunityTime User defined time to avoid being detected early.
+     * @return Return true if put success, otherwise return false.
+     */
+    private boolean checkPrepareQueueOffset(HashMap<Long, Long> removeMap, List<Long> doneOpOffset, MessageExt msgExt,
+        long checkImmunityTime) {
+        if (System.currentTimeMillis() - msgExt.getBornTimestamp() < checkImmunityTime) {
+            String prepareQueueOffsetStr = msgExt.getUserProperty(MessageConst.PROPERTY_TRANSACTION_PREPARED_QUEUE_OFFSET);
+            if (null == prepareQueueOffsetStr) {
+                return putImmunityMsgBackToHalfQueue(msgExt);
+            } else {
+                long prepareQueueOffset = getLong(prepareQueueOffsetStr);
+                if (-1 == prepareQueueOffset) {
+                    return false;
+                } else {
+                    if (removeMap.containsKey(prepareQueueOffset)) {
+                        long tmpOpOffset = removeMap.remove(prepareQueueOffset);
+                        doneOpOffset.add(tmpOpOffset);
+                        return true;
+                    } else {
+                        return putImmunityMsgBackToHalfQueue(msgExt);
+                    }
+                }
+
+            }
+
+        } else {
+            return true;
+        }
+    }
+
+    /**
+     * Write messageExt to Half topic again
+     *
+     * @param messageExt Message will be write back to queue
+     * @return Put result can used to determine the specific results of storage.
+     */
+    private PutMessageResult putBackToHalfQueueReturnResult(MessageExt messageExt) {
+        PutMessageResult putMessageResult = null;
+        try {
+            MessageExtBrokerInner msgInner = transactionBridge.renewHalfMessageInner(messageExt);
+            putMessageResult = transactionBridge.putMessageReturnResult(msgInner);
+        } catch (Exception e) {
+            e.printStackTrace();
```

Comment:
Just printStackTrace?
We need to recover from the error and mark tx service unavailable and propagate the unavailable state to producer clients if recovering is not possible.

Reply:
Thanks for helping to point out the problem，output to the log will be a good choice, as in this place, The outer method uses this method to decide whether to send back the request ，based on the status of putMessageResult, so there's no need to propagate the unavailable state to producer.